votor-messages/src/vote.rs

```diff
@@ -0,0 +1,263 @@
+        matches!(self, Self::Notarize(_))
+    }
+
+    /// Whether the vote is a finalization vote
+    pub fn is_finalize(&self) -> bool {
+        matches!(self, Self::Finalize(_))
+    }
+
+    /// Whether the vote is a skip vote
+    pub fn is_skip(&self) -> bool {
+        matches!(self, Self::Skip(_))
+    }
+
+    /// Whether the vote is a notarization fallback vote
+    pub fn is_notarize_fallback(&self) -> bool {
+        matches!(self, Self::NotarizeFallback(_))
+    }
+
+    /// Whether the vote is a skip fallback vote
+    pub fn is_skip_fallback(&self) -> bool {
+        matches!(self, Self::SkipFallback(_))
+    }
+
+    /// Whether the vote is a notarization or finalization
+    pub fn is_notarization_or_finalization(&self) -> bool {
+        matches!(self, Self::Notarize(_) | Self::Finalize(_))
+    }
+}
+
+impl From<NotarizationVote> for Vote {
+    fn from(vote: NotarizationVote) -> Self {
+        Self::Notarize(vote)
+    }
+}
+
+impl From<FinalizationVote> for Vote {
+    fn from(vote: FinalizationVote) -> Self {
+        Self::Finalize(vote)
+    }
+}
+
+impl From<SkipVote> for Vote {
+    fn from(vote: SkipVote) -> Self {
+        Self::Skip(vote)
+    }
+}
+
+impl From<NotarizationFallbackVote> for Vote {
+    fn from(vote: NotarizationFallbackVote) -> Self {
+        Self::NotarizeFallback(vote)
+    }
+}
+
+impl From<SkipFallbackVote> for Vote {
+    fn from(vote: SkipFallbackVote) -> Self {
+        Self::SkipFallback(vote)
+    }
+}
+
+/// A notarization vote
+#[cfg_attr(
+    feature = "frozen-abi",
+    derive(AbiExample),
+    frozen_abi(digest = "5AdwChAjsj5QUXLdpDnGGK2L2nA8y8EajVXi6jsmTv1m")
+)]
+#[derive(Clone, Copy, Debug, PartialEq, Default, Serialize, Deserialize)]
+pub struct NotarizationVote {
+    slot: Slot,
+    block_id: Hash,
+}
+
+impl NotarizationVote {
+    /// Construct a notarization vote for `slot`
+    pub fn new(slot: Slot, block_id: Hash) -> Self {
+        Self { slot, block_id }
+    }
+
+    /// The slot to notarize
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+
+    /// The block_id of the notarization slot
+    pub fn block_id(&self) -> &Hash {
+        &self.block_id
+    }
+}
+
+/// A finalization vote
+#[cfg_attr(
+    feature = "frozen-abi",
+    derive(AbiExample),
+    frozen_abi(digest = "2XQ5N6YLJjF28w7cMFFUQ9SDgKuf9JpJNtAiXSPA8vR2")
+)]
+#[derive(Clone, Copy, Debug, PartialEq, Default, Serialize, Deserialize)]
+pub struct FinalizationVote {
+    slot: Slot,
+}
+
+impl FinalizationVote {
+    /// Construct a finalization vote for `slot`
+    pub fn new(slot: Slot) -> Self {
+        Self { slot }
+    }
+
+    /// The slot to finalize
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+}
+
+/// A skip vote
+/// Represents a range of slots to skip
+/// inclusive on both ends
+#[cfg_attr(
+    feature = "frozen-abi",
+    derive(AbiExample),
+    frozen_abi(digest = "G8Nrx3sMYdnLpHsCNark3BGA58BmW2sqNnqjkYhQHtN")
+)]
```

Comment:
If we need frozen-abi, maybe we should do it for ConsensusMessage as well, otherwise let's remove all of them.

Reply:
can address separately but it might be a good idea to version ConsensusMessage for future proofing

Reply:
@alexpyattaev is submitting TLV crate to do versioning https://github.com/anza-xyz/agave/pull/7694/files/43d7d744ad59158b59d9efd209a501c2eff4bd14..479b657b9d5fea4ccc2196eebc719cf470a296f9

Reply:
Yes please! Gossip is a mess because it is essentially impossible to deprecate stuff from parser without breaking abi.

Reply:
btw, I'm happy to use this PR as a landing spot for this versioning design discussion (i.e. can up-level from precise code semantics for a second). Don't need to approve the PR w/ haste.

Current version:

1. keeps ABI versioning
2. renames `ConsensusMessage` enum --> `Message`
3. Introduces `ConsensusMessage` struct that wraps `Message` and has major/minor version fields
4. Test to simulate some future unknown message that passes/fails deserialization

Some questions:

1. Do we need/want to freeze the ABI if we have versioning?
2. Are we aligned on just versioning the top level struct?
3. How big of sticklers do we want to be on rolling version for breaking changes in the near future (before we're running any of this on persistent clusters)?

Reply:
I'd prefer keeping the ABI just because it creates a test that will fail if the format is accidentally changed. AFAIK it doesn't add any overhead, when we add a new version we can update the digest.

And I'm inclined to just not version for now and once TLV lands update this rather than hand rolling our own versioning scheme. Since we're not on testnet yet we can play loose with format updates, but we'll ensure we have versioning before launch.

Reply:
Versioning with current impl is better than no versioning, but I believe your Message field should be Vec<u8> and then deserialized if parser for relevant version is available. See also comment on the test you have added. To avoid double deserialization costs, TLV crate can be used. One can trivially enforce that all fields are parsed in TLV packet if desired.

Reply:
I find myself agreeing with Ashwin:

Keep the ABI digest as a hard reminder that when we change things here, we MUST update the parsers and versions and such. I can add a comment to that effect.

I'm leaning towards removing the naive versioning so that we:

1. Don't dupe ourselves into thinking we have a proper solution here
2. Don't invest resources in putting lipstick on this pig